### PR TITLE
Haby and Lynn Scrabble.rb

### PR DESCRIPTION
**Pull request by habypsow**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	| We feel good! We did a good job switching back and forth after each 'challenge' of the code, and had good communication throughout.  We also started out by both pseudo coding each wave, then comparing, and creating a skeleton structure to work from.  This kept us on track of what each next step would look like.	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |Because we did not want to create a new object/instance for every word that we were scoring. A class method made more sense here.    |
|   Describe an example of a test you wrote for a _nominal case_. |For example, we wanted to ensure that our #total_score method returned the sum of all the words played. So we tested dummy numbers in the parameters of the method, to ensure that the output would be the expected sum of all the words played.   	|
|   Describe an example of a test you wrote for an _edge case_. |For example, we  wanted to ensure that the input for word would be a 'string', in our #Play specs, and so we raised an argument error if the input was anything other than a string.   	|
|   What was your final test coverage percentage?	| 100% :1st_place_medal:  :)  	|
|   Is there a specific piece of code you'd like feedback on?	|We wrote self.highest_score a while back. Looking over the entire code, we are wondering if there would be a way to make this specific method drier, maybe through the use of enumerables?   	|

### CODE REVIEW

File: lib/scrabble_scoring.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    def self.highest_score_from(array_of_words)
+      raise ArgumentError.new("Argument must be an array") if array_of_words.class != Array
+
+      word_scores = {}
+      array_of_words.each do |word|
+        word_scores[word] = Scrabble::Scoring.score(word)
+      end
+
+      winner = []
+      word_scores.each do |word, score|
```
Comment (CheezItMan) [inline lib/scrabble_scoring.rb:68]: You could use `.select` here to DRY it up a bit.

File: lib/scrabble_scoring.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      end
+
+      return score_total
+    end
+
+    def self.highest_score_from(array_of_words)
+      raise ArgumentError.new("Argument must be an array") if array_of_words.class != Array
+
+      word_scores = {}
+      array_of_words.each do |word|
```
Comment (CheezItMan) [inline lib/scrabble_scoring.rb:63]: This loop is probably a good candidate for `map`

File: lib/scrabble_scoring.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        if score == word_scores.values.max
+          winner << word
+        end
+      end
+
+      winner.each do |winning_word|
+        if winning_word.length == 7
+          return winning_word
+        else
+          return winner.min_by {|word| word.length}
```
Comment (CheezItMan) [inline lib/scrabble_scoring.rb:78]: Good use of `min_by`

File: lib/scrabble_scoring.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      end
+
+      winner = []
+      word_scores.each do |word, score|
+        if score == word_scores.values.max
+          winner << word
+        end
+      end
+
+      winner.each do |winning_word|
```
Comment (CheezItMan) [inline lib/scrabble_scoring.rb:74]: `select` could have been used to select words that are length 7.

File: lib/scrabble_scoring.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    def self.score(word)
+      if word =~ /\W/
+        raise ArgumentError.new "Please enter a real word!"
+      end
+      score_total = 0
+      word_array = word.upcase.split("")
+
+      word_array.each do |letter|
```
Comment (CheezItMan) [inline lib/scrabble_scoring.rb:48]: `map` could be useful here, but well done.

File: lib/scrabble_scoring.rb [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        score_total += SCORE_CHART[letter]
+      end
+
+      if word.length >= 7
```
Comment (CheezItMan) [inline lib/scrabble_scoring.rb:52]: I believe in Scrabble words have a max length of 7, so words that are longer shouldn't be allowed.

File: specs/scrabble_scoring_spec.rb [hunk 7]
```diff
@@ -0,0 +1,59 @@
+require_relative '../specs/spec_helper.rb'
+require_relative '../lib/scrabble_scoring'
+
+describe "Scoring Class" do
+
+  describe "Scoring is a class" do
+    before do
+      @my_score = Scrabble::Scoring.new("Happy")
```
Comment (CheezItMan) [inline specs/scrabble_scoring_spec.rb:8]: Since we only have class methods, you don't need to create an instance.

File: specs/scrabble_scoring_spec.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    it "Word cannot include symbols or numbers" do
+      proc {Scrabble::Scoring.score("@&89")}.must_raise ArgumentError
+    end
+
+    it "7 letter word receives a 50 point bonus" do
+      word = Scrabble::Scoring.score("Scrabble")
+      word.must_equal 64
+    end
+  end
```
Comment (CheezItMan) [inline specs/scrabble_scoring_spec.rb:35]: You'll need to also test, an empty string (""), as well.

File: specs/scrabble_scoring_spec.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+  describe "self.highest_score_from" do
+
+    it "Raise an ArgumentError if parameter entered isn't an Array" do
+      proc { Scrabble::Scoring.highest_score_from("123")}.must_raise ArgumentError
+    end
+
+    it "returns the word in the array with the highest score" do
```
Comment (CheezItMan) [inline specs/scrabble_scoring_spec.rb:43]: What about testing an empty array?

File: specs/scrabble_scoring_spec.rb [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    end
+
+    it "returns the word with the fewest letters if a tie occurs" do
+      tie_array = ["que", "jaaaa", "dope"]
+      Scrabble::Scoring.highest_score_from(tie_array).must_equal "que"
+    end
+
+    it "returns the word with 7 letters when tied with top score" do
+      winning_words = ["liberty", "xaaaa"]
+      Scrabble::Scoring.highest_score_from(winning_words).must_equal "liberty"
```
Comment (CheezItMan) [inline specs/scrabble_scoring_spec.rb:56]: It's also good to test, in case of a tie that the right answer is selected regardless of the order.

File: specs/scrabble_player_spec.rb [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    it "Scrabble::Player(name) must be a String" do
+      proc {
+        Scrabble::Player.new(123)}.must_raise ArgumentError
+      end
+    end
+
+  describe "Play" do
+    it "Word must be a string" do
+      proc { @new_player.play(123) }.must_raise ArgumentError
+    end
```
Comment (CheezItMan) [inline specs/scrabble_player_spec.rb:23]: You should also test an emptystring

File: lib/scrabble_tilebag.rb [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      raise ArgumentError.new("Please draw less than 7 tiles") if num > 7
+      new_tiles = []
+      num.times {new_tiles << @tiles_remaining.pop }
+      return new_tiles
+      #also may eventually update tiles in player hand
+    end
+
+    def tiles_remaining
+      if @tiles_remaining.empty?
+        puts "There are no more tiles!"
```
Comment (CheezItMan) [inline lib/scrabble_tilebag.rb:38]: You should not have a puts message here!  If it's empty you should return 0.

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|   Good work, I left some notes in the PR suggesting some methods to DRY things up, but you did really well	|
|   Both Teammates contributed to the codebase	|   Good work collaborating	|
|   Regular Commits with meaningful commit messages 	|   Nice regular commits and decent messages	|
|  **Scoring class**	|   	|
|  `score` method	|   Well done	|
|   Uses appropriate data structure to store the letter score	|   Good use of a hash!	|
|   Appropriately handles edge cases	|   nice use of regex	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|  You should also test an empty string. 	|
|  `highest_scoring_word` method	|   Check	|
|   Appropriately handles edge cases	|   Missing a test case for an empty array	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|   Missing tests for an empty array, and tests which vary the order of the words	|
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|   check	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   Check	|
|  Returns `true` or `false` from `won?` method | Good use of a ternary operator  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |  Similar feedback to testing in scoring.  Otherwise quite good.   |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|   Good use of the hash, array & `shuffle!` to set up the Tilebag!	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |   Check, well done	|
| `tiles_remaining` method returns a count	|   You should not have a puts message in  `tiles_remaining` !  If it's empty you should return 0.	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   Good work testing if you can draw too many tiles, Also good work testing the nominal case	|


## Summary

Well done.  You've written some pretty good tests and did a good job using enumerables where you could.  I put a few notes in the PR code to identify some places where you could DRY up your code a bit more and a few places where you missed a few things.

File: specs/scrabble_player_spec.rb [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+    it "object.play(word) would return correct score as a fixnum"  do
+      score = @new_player.play("hi")
+      score.must_equal 5
+      score.must_be_instance_of Fixnum
```
Comment (CheezItMan) [inline specs/scrabble_player_spec.rb:28]: As of Ruby 2.4.0, this should be:

`score.must_be_instance_of Integer`

**Review by CheezItMan — COMMENTED** [review]
Just a few comments in the code to go along with the assessment.
